nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/upgrade/tasks/P2RewriteCompositeMetadataTask.java

```diff
@@ -0,0 +1,235 @@
+import org.sonatype.nexus.repository.storage.BucketStore;
+import org.sonatype.nexus.repository.storage.StorageFacet;
+import org.sonatype.nexus.repository.storage.StorageTx;
+import org.sonatype.nexus.repository.storage.TempBlob;
+import org.sonatype.nexus.repository.transaction.TransactionalStoreBlob;
+import org.sonatype.nexus.repository.transaction.TransactionalTouchBlob;
+import org.sonatype.nexus.repository.types.ProxyType;
+import org.sonatype.nexus.repository.view.Content;
+import org.sonatype.nexus.scheduling.Cancelable;
+import org.sonatype.nexus.scheduling.CancelableHelper;
+import org.sonatype.nexus.scheduling.TaskConfiguration;
+import org.sonatype.nexus.scheduling.TaskScheduler;
+import org.sonatype.nexus.scheduling.TaskSupport;
+import org.sonatype.nexus.transaction.Transactional;
+import org.sonatype.nexus.transaction.UnitOfWork;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.commons.io.IOUtils;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Streams.stream;
+import static org.sonatype.nexus.repository.p2.internal.P2FacetImpl.HASH_ALGORITHMS;
+import static org.sonatype.nexus.repository.p2.upgrade.P2Upgrade_1_2.MARKER_FILE;
+
+/**
+ * Rewrites composite metadata to reference site hashes rather than URL paths.
+ *
+ * @since 1.next
+ */
+@Named
+public class P2RewriteCompositeMetadataTask
+    extends TaskSupport
+    implements Cancelable
+{
+  private final Path markerFile;
+
+  private final RepositoryManager repositoryManager;
+
+  private final BucketStore bucketStore;
+
+  private final ContentValidator contentValidator;
+
+  private final TaskScheduler taskScheduler;
+
+  @Inject
+  public P2RewriteCompositeMetadataTask(
+      final ApplicationDirectories directories,
+      final BucketStore bucketStore,
+      final RepositoryManager repositoryManager,
+      final TaskScheduler taskScheduler,
+      final ContentValidator contentValidator)
+  {
+    markerFile = new File(directories.getWorkDirectory("db"), MARKER_FILE).toPath();
+    this.bucketStore = checkNotNull(bucketStore);
+    this.contentValidator = checkNotNull(contentValidator);
+    this.repositoryManager = checkNotNull(repositoryManager);
+    this.taskScheduler = checkNotNull(taskScheduler);
+  }
+
+  @Override
+  protected Object execute() throws Exception {
+    List<Repository> repositories = stream(repositoryManager.browse())
+        .peek(r -> log.debug("Looking at repository: {}", r))
+        .filter(r -> r.getFormat() instanceof P2Format)
+        .peek(r -> log.debug("Looking at p2 repository: {}", r))
+        .filter(r -> r.getType() instanceof ProxyType)
+        .peek(r -> log.debug("Found p2 proxy repository: {}", r))
+        .collect(Collectors.toList());
+
+    if (!repositories.isEmpty()) {
+      Iterable<Asset> compositeAssets = rewriteLegacyAssets(repositories);
+      CancelableHelper.checkCancellation();
+
+      compositeAssets.forEach(asset -> {
```

Comment:
Question: `compositeAssets` holds list of all assets of all repositories and they will process all of them in one go. Wouldn't it make sense to process assets repository by repository with `checkCancellation()` between them?

Reply:
> Question: `compositeAssets` holds list of all assets of all repositories and they will process all of them in one go. Wouldn't it make sense to process assets repository by repository with `checkCancellation()` between them?

My thought was the number would be limited enough, and the work per asset small enough that it wouldn't matter.